### PR TITLE
Fixes deeplearning4j/nd4j#1859

### PR DESCRIPTION
**Pull request by jarney**
## What changes were proposed in this pull request?
See issue [#1859](https://github.com/deeplearning4j/nd4j/issues/1859)

The use of an MSE or L2 cost function means that the network will effectively converge on the mean of the dataset. For some datasets, this is not a desirable behavior and instead, one would like to find the parameters of the distribution of the data. In particular, it is often desirable to converge on a mixture of gaussian distributions which describe the statistical properties of the data.

For this purpose, nd4j should provide a Mixture Density loss function as described by [Bishop1994](
http://publications.aston.ac.uk/373/1/NCRG_94_004.pdf)
## How was this patch tested?

o.d.g.LossFunctionGradientCheck -  ***** Starting test: LossMixtureDensity(mixtures=2, labels=3) - identity - minibatchSize = 1 *****
o.d.g.GradientCheckUtil - Param 0 (0_W) passed: grad= -0.6053853982355434, numericalGrad= -0.6053853986998092, relError= 3.834465194925549E-10
o.d.g.GradientCheckUtil - Param 1 (0_W) passed: grad= -0.47783684344684657, numericalGrad= -0.4778368429825264, relError= 4.858564173090874E-10
o.d.g.GradientCheckUtil - Param 2 (0_W) passed: grad= -0.13606158295825885, numericalGrad= -0.13606158244527933, relError= 1.8851005176736425E-9
...
o.d.g.GradientCheckUtil - Param 69 (1_b) passed: grad= -0.04692207396946135, numericalGrad= -0.04692207333789611, relError= 6.729937483240737E-9
o.d.g.GradientCheckUtil - GradientCheckUtil.checkGradients(): 70 params checked, 70 passed, 0 failed. Largest relative error = 8.280817397437084E-4

o.d.g.LossFunctionGradientCheck -  ***** Starting test: LossMixtureDensity(mixtures=2, labels=3) - identity - minibatchSize = 3 *****
o.d.g.GradientCheckUtil - Param 0 (0_W) passed: grad= -0.2508574880968211, numericalGrad= -0.2508574881154857, relError= 3.720154026603924E-11
o.d.g.GradientCheckUtil - Param 1 (0_W) passed: grad= -0.22399699737871923, numericalGrad= -0.2239969978390377, relError= 1.0275103744545979E-9
...
o.d.g.GradientCheckUtil - Param 68 (1_b) passed: grad= -0.06625373683265866, numericalGrad= -0.06625373627500153, relError= 4.208495685031032E-9
o.d.g.GradientCheckUtil - Param 69 (1_b) passed: grad= 0.026636581482483768, numericalGrad= 0.026636581829109218, relError= 6.5065678292319444E-9
o.d.g.GradientCheckUtil - GradientCheckUtil.checkGradients(): 70 params checked, 70 passed, 0 failed. Largest relative error = 3.088100049488012E-8

o.d.g.LossFunctionGradientCheck -  ***** Starting test: LossMixtureDensity(mixtures=2, labels=3) - tanh - minibatchSize = 1 *****
o.d.g.GradientCheckUtil - Param 0 (0_W) passed: grad= -0.01937115689025789, numericalGrad= -0.019371156678715806, relError= 5.46023360441808E-9
o.d.g.GradientCheckUtil - Param 1 (0_W) passed: grad= -0.015289850877362983, numericalGrad= -0.015289850718858133, relError= 5.183335402132627E-9
...
o.d.g.GradientCheckUtil - Param 68 (1_b) passed: grad= -0.10644004203090246, numericalGrad= -0.10644004211357583, relError= 3.8835651763115025E-10
o.d.g.GradientCheckUtil - Param 69 (1_b) passed: grad= -0.12969292609902738, numericalGrad= -0.12969292617626138, relError= 2.977571687052128E-10
o.d.g.GradientCheckUtil - GradientCheckUtil.checkGradients(): 70 params checked, 70 passed, 0 failed. Largest relative error = 2.3422202304370755E-6

o.d.g.LossFunctionGradientCheck -  ***** Starting test: LossMixtureDensity(mixtures=2, labels=3) - tanh - minibatchSize = 3 *****
o.d.g.GradientCheckUtil - Param 0 (0_W) passed: grad= -0.021537849488570588, numericalGrad= -0.021537849281827448, relError= 4.7995307300946E-9
o.d.g.GradientCheckUtil - Param 1 (0_W) passed: grad= -0.02791281404883191, numericalGrad= -0.027912813838071315, relError= 3.775337617925421E-9
o.d.g.GradientCheckUtil - Param 2 (0_W) passed: grad= -0.060868752629540133, numericalGrad= -0.060868752704834606, relError= 6.184985624546654E-10
...
o.d.g.GradientCheckUtil - Param 68 (1_b) passed: grad= -0.0682741795109804, numericalGrad= -0.0682741791990793, relError= 2.2841805285277727E-9
o.d.g.GradientCheckUtil - Param 69 (1_b) passed: grad= -0.04039526858033265, numericalGrad= -0.04039526846355557, relError= 1.4454301869187682E-9
o.d.g.GradientCheckUtil - GradientCheckUtil.checkGradients(): 70 params checked, 70 passed, 0 failed. Largest relative error = 3.4317726044166616E-7

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Reviewed the [Contributing Guidelines](https://github.com/deeplearning4j/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X ] Created tests for any significant new code additions.
- [ X ] Relevant tests for your changes are passing.
- [ X ] Ran mvn formatter:format (see [formatter instructions](http://code.revelc.net/formatter-maven-plugin/examples.html#Setting_Source_Files) for targeting your specific files).

### CODE REVIEW

**Comment by jarney** [comment]
Re-based changes to latest master.  All tests still pass, including ones formerly in conflict.
---- Passed ----
LossBinaryXENT() - sigmoid - minibatchSize = 1
LossBinaryXENT() - sigmoid - minibatchSize = 3
LossBinaryXENT() - sigmoid - minibatchSize = 1
LossBinaryXENT() - sigmoid - minibatchSize = 3
LossCosineProximity() - tanh - minibatchSize = 1
LossCosineProximity() - tanh - minibatchSize = 3
LossHinge() - tanh - minibatchSize = 1
LossHinge() - tanh - minibatchSize = 3
LossKLD() - sigmoid - minibatchSize = 1
LossKLD() - sigmoid - minibatchSize = 3
LossKLD() - softmax - minibatchSize = 1
LossKLD() - softmax - minibatchSize = 3
LossL1() - tanh - minibatchSize = 1
LossL1() - tanh - minibatchSize = 3
LossL1() - rationaltanh - minibatchSize = 1
LossL1() - rationaltanh - minibatchSize = 3
LossL1() - softmax - minibatchSize = 1
LossL1() - softmax - minibatchSize = 3
LossL2() - tanh - minibatchSize = 1
LossL2() - tanh - minibatchSize = 3
LossL2() - softmax - minibatchSize = 1
LossL2() - softmax - minibatchSize = 3
LossMAE() - identity - minibatchSize = 1
LossMAE() - identity - minibatchSize = 3
LossMAE() - softmax - minibatchSize = 1
LossMAE() - softmax - minibatchSize = 3
LossMAPE() - identity - minibatchSize = 1
LossMAPE() - identity - minibatchSize = 3
LossMAPE() - softmax - minibatchSize = 1
LossMAPE() - softmax - minibatchSize = 3
LossMCXENT() - softmax - minibatchSize = 1
LossMCXENT() - softmax - minibatchSize = 3
LossMSE() - identity - minibatchSize = 1
LossMSE() - identity - minibatchSize = 3
LossMSE() - softmax - minibatchSize = 1
LossMSE() - softmax - minibatchSize = 3
LossMSLE() - sigmoid - minibatchSize = 1
LossMSLE() - sigmoid - minibatchSize = 3
LossMSLE() - softmax - minibatchSize = 1
LossMSLE() - softmax - minibatchSize = 3
LossNegativeLogLikelihood() - sigmoid - minibatchSize = 1
LossNegativeLogLikelihood() - sigmoid - minibatchSize = 3
LossNegativeLogLikelihood() - softmax - minibatchSize = 1
LossNegativeLogLikelihood() - softmax - minibatchSize = 3
LossPoisson() - sigmoid - minibatchSize = 1
LossPoisson() - sigmoid - minibatchSize = 3
LossSquaredHinge() - tanh - minibatchSize = 1
LossSquaredHinge() - tanh - minibatchSize = 3
LossFMeasure(beta=1.0) - sigmoid - minibatchSize = 1
LossFMeasure(beta=1.0) - sigmoid - minibatchSize = 3
LossFMeasure(beta=2.0) - sigmoid - minibatchSize = 1
LossFMeasure(beta=2.0) - sigmoid - minibatchSize = 3
LossFMeasure(beta=1.0) - softmax - minibatchSize = 1
LossFMeasure(beta=1.0) - softmax - minibatchSize = 3
LossFMeasure(beta=2.0) - softmax - minibatchSize = 1
LossFMeasure(beta=2.0) - softmax - minibatchSize = 3
LossMixtureDensity(mixtures=2, labels=3) - identity - minibatchSize = 1
LossMixtureDensity(mixtures=2, labels=3) - identity - minibatchSize = 3
LossMixtureDensity(mixtures=2, labels=3) - tanh - minibatchSize = 1
LossMixtureDensity(mixtures=2, labels=3) - tanh - minibatchSize = 3

**Comment by jarney** [comment]
Also added mask test to ensure that masking works in the same way as other loss functions:

testPerOutputMaskingMLP(): maskShape = [1, 6], loss function = LossMixtureDensity(mixtures=2, labels=1), activation = TANH
o.d.g.GradientCheckUtil - Param 0 (0_W) passed: grad= -0.05927297483473185, numericalGrad= -0.059272974861634964, relError= 2.2694247885954322E-10
o.d.g.GradientCheckUtil - Param 1 (0_W) passed: grad= -0.04678476104524873, numericalGrad= -0.04678476106523277, relError= 2.1357426807451713E-10
o.d.g.GradientCheckUtil - Param 2 (0_W) passed: grad= -0.013321720025233932, numericalGrad= -0.013321720060410769, relError= 1.3202813534788304E-9
o.d.g.GradientCheckUtil - Param 3 (0_W) passed: grad= -1.5770338653776933E-4, numericalGrad= -1.5770340588971976E-4, relError= 6.135552847792436E-8
o.d.g.GradientCheckUtil - Param 4 (0_W) passed: grad= -0.05182378004831799, numericalGrad= -0.05182378004509047, relError= 3.1139329031877915E-11
o.d.g.GradientCheckUtil - Param 5 (0_W) passed: grad= -0.10618585853343684, numericalGrad= -0.10618585866151165, relError= 6.030690077589474E-10
o.d.g.GradientCheckUtil - Param 6 (0_W) passed: grad= 0.00207674179131346, numericalGrad= 0.002076741667522697, relError= 2.980408239426757E-8
o.d.g.GradientCheckUtil - Param 7 (0_W) passed: grad= 0.0016391933883897086, numericalGrad= 0.0016391934476445158, relError= 1.8074379317536742E-8
o.d.g.GradientCheckUtil - Param 8 (0_W) passed: grad= 4.6675188457673885E-4, numericalGrad= 4.667517483625261E-4, relError= 1.4591717477509385E-7
o.d.g.GradientCheckUtil - Param 9 (0_W) passed: grad= 5.525439112307509E-6, numericalGrad= 5.525357948954479E-6, relError= 7.344570041381958E-6
o.d.g.GradientCheckUtil - Param 10 (0_W) passed: grad= 0.001815745035747969, numericalGrad= 0.0018157449988009944, relError= 1.0174053682692809E-8
o.d.g.GradientCheckUtil - Param 11 (0_W) passed: grad= 0.0037204242013793793, numericalGrad= 0.0037204243019672845, relError= 1.3518338015811312E-8
o.d.g.GradientCheckUtil - Param 12 (0_W) passed: grad= 0.020959593657077384, numericalGrad= 0.020959593727276626, relError= 1.6746327071857826E-9
o.d.g.GradientCheckUtil - Param 13 (0_W) passed: grad= 0.016543620150431287, numericalGrad= 0.016543620251496804, relError= 3.054516358531251E-9
o.d.g.GradientCheckUtil - Param 14 (0_W) passed: grad= 0.0047107107298188485, numericalGrad= 0.004710710821420605, relError= 9.722710746327882E-9
o.d.g.GradientCheckUtil - Param 15 (0_W) passed: grad= 5.576569945060032E-5, numericalGrad= 5.5765725370804375E-5, relError= 2.32402698112107E-7

**Comment by jarney** [comment]
@AlexDBlack : LossMixtureDensity now participates in gradientCheckMaskingOutputSimple() test, so the mask is tested/exercised.  I think that closes the last issue.
